go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -576,16 +587,28 @@ func egressGetL4Match(ports []egressfirewallapi.EgressFirewallPort) string {
 	return fmt.Sprintf("(%s)", l4Match)
 }
 
-func getClusterSubnetsExclusion() string {
+func getClusterSubnetsExclusion(destinations []matchTarget) string {
 	var exclusion string
-	for _, clusterSubnet := range config.Default.ClusterSubnets {
-		if exclusion != "" {
-			exclusion += " && "
-		}
-		if utilnet.IsIPv6CIDR(clusterSubnet.CIDR) {
-			exclusion += fmt.Sprintf("%s.dst != %s", "ip6", clusterSubnet.CIDR)
-		} else {
-			exclusion += fmt.Sprintf("%s.dst != %s", "ip4", clusterSubnet.CIDR)
+	for _, destination := range destinations {
```

Comment:
Yes, indeed. That seems wrong to me too, we need `ipv4.dst != || ipv6.dst != ?` to make it work.

Reply:
Actually no, I believe this is right - I will try testing it now though. This func `getClusterSubnetsExclusion` is supposed to generate a match condition which excludes east<->west traffic and does not have it blocked by egress firewall. If we have `subnet1 = 10.243.0.0/16` and `subnet2 = 10.245.0.0/16` then we'll need an exclusion filter similar to `ip4.dst != 10.243.0.0/16 && ip4.dst != 10.245.0.0/16`. Conversely, for dual-stack we might be fine with `ip4.dst != 10.243.0.0/16 && ip6.dst != ffff:0af3:0000::/64`, depending on how OVN/OVS defines the packet flow for a non-defined IP protocol. I.e if we have a IPv6 connection to an IPv6 destination, what will the default value of `ip4.dst` be? If the answer is "anything except an error, leading to an un-parsable action" then we're fine, if it's an error this will fail.

Reply:
OK, the answer to the question above is: it's "an error". We thus need the `||` operator. If we sent traffic to an IPv6 address, the filter `ip4.dst !=  10.244.0.0/16` does not return true, it returns false.

Reply:
Actually, the answer to 

> isnt it possible in dualstack that we build up an exclusion string

Is no. This is because the exclusion filter will match the IP family of the subnets defined with the specified destination IP in the egress firewall rule. This means that we won't combine a "deny 0.0.0.0/0" with an IPv6 subnet. The user will need to define a "deny ::/0" for the IPv6 subnet range.

Reply:
maybe i read the code wrong, but it looked like we iterate over all destinations, which right now are hardcoded to pass unknownIPv4 and ipv6. Therefore we will iterate through both. In the case where we have dualstack, I think we will end up with exclusion criteria for both. My point wasn't whether or not it should be && or ||, it was really about is that even valid/legal match criteria? I wonder if OVN you can actually combine ipv4 and ipv6. In openflow you cant so I doubt that is allowed in OVN either.

Reply:
Yes, it's as you describe it, however the `||` operators can give the illusion that it works. As I mentioned above: if you do `ip4.dst != 10.243.0.0/16 || ip6.dst != ffff:0af3:0000::/64` and send traffic to `ffff:0af3:0000:6` the `ip4.dst` part is "error", which ends up having the entire statement evaluate to what you desired because of the `||` operator. But it's technically wrong.

Reply:
It seems ovn-controller avoids writing flows which combine the IPv4 and IPv6 match criteria, see following output:
```
 ovn-nbctl list acl 6683e521-bce1-4fa6-a335-994dcaae1f61
_uuid               : 6683e521-bce1-4fa6-a335-994dcaae1f61
action              : drop
direction           : to-lport
external_ids        : {egressFirewall=egressip-8081-blockAll}
log                 : false
match               : "(ip4.dst == 0.0.0.0/0 || ip6.dst == ::/0) && ip4.src == $a12305685082940126680 && ip4.dst != 10.244.0.0/16"
meter               : []
name                : []
priority            : 10000
severity            : []
[root@ovn-control-plane ~]# ovn-sbctl dump-flows | grep 11000
  table=4 (ls_out_acl         ), priority=11000, match=(reg0[10] == 1 && ((ip4.dst == 0.0.0.0/0 || ip6.dst == ::/0) && ip4.src == $a12305685082940126680 && ip4.dst != 10.244.0.0/16)), action=(ct_commit { ct_label.blocked = 1; }; /* drop */)
  table=4 (ls_out_acl         ), priority=11000, match=(reg0[9] == 1 && ((ip4.dst == 0.0.0.0/0 || ip6.dst == ::/0) && ip4.src == $a12305685082940126680 && ip4.dst != 10.244.0.0/16)), action=(/* drop */)
[root@ovn-control-plane ~]# ovs-ofctl -O OpenFlow13 dump-flows br-int | grep -i "priority=11000"
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_src=10.244.1.5 actions=conjunction(24,2/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_src=10.244.0.3 actions=conjunction(24,2/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=16.0.0.0/16.0.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=128.0.0.0/1 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/0.64.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=64.0.0.0/64.0.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=1.0.0.0/1.0.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.1.0.0/0.1.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/8.0.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.905s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=32.0.0.0/32.0.0.0 actions=conjunction(24,1/2)
 cookie=0x762c2740, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,conj_id=24,ip,reg0=0x200/0x200,metadata=0x4 actions=drop
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=4.0.0.0/4.0.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.2.0.0/0.2.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/2.0.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/0.32.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/0.4.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/0.16.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.8.0.0/0.8.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x200/0x200,metadata=0x4,nw_dst=0.0.0.0/0.128.0.0 actions=conjunction(24,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/0.32.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=32.0.0.0/32.0.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.8.0.0/0.8.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.2.0.0/0.2.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=4.0.0.0/4.0.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/0.64.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=1.0.0.0/1.0.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=128.0.0.0/1 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/0.128.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=16.0.0.0/16.0.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/2.0.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.1.0.0/0.1.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_src=10.244.1.5 actions=conjunction(25,2/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_src=10.244.0.3 actions=conjunction(25,2/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/0.4.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/0.16.0.0 actions=conjunction(25,1/2)
 cookie=0xea40bcd4, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,conj_id=25,ip,reg0=0x400/0x400,metadata=0x4 actions=ct(commit,zone=NXM_NX_REG13[0..15],exec(load:0x1->NXM_NX_CT_LABEL[0]))
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=0.0.0.0/8.0.0.0 actions=conjunction(25,1/2)
 cookie=0x0, duration=648.904s, table=44, n_packets=0, n_bytes=0, priority=11000,ip,reg0=0x400/0x400,metadata=0x4,nw_dst=64.0.0.0/64.0.0.0 actions=conjunction(25,1/2)
```

There are quite a few flows created, but again, that's because of the `!=` operator.